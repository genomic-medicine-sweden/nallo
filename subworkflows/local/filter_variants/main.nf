include { ENSEMBLVEP_FILTERVEP          } from '../../../modules/nf-core/ensemblvep/filtervep/main'
include { BCFTOOLS_VIEW                 } from '../../../modules/nf-core/bcftools/view/main'
include { ENSEMBLVEP_VEP } from '../../../modules/nf-core/ensemblvep/vep/main.nf'

workflow FILTER_VARIANTS {

    take:
    ch_vcf      // channel  [optional] [ val(meta), path(vcf) ]
    ch_hgnc_ids // channel: [optional] [ val(meta), path(txt) ]
    filter_hgnc //    bool: should filter_vep be run to filter on hgnc ids

    main:
    ch_versions = Channel.empty()

    if ( filter_hgnc ) {

        ENSEMBLVEP_FILTERVEP (
            ch_vcf,
            ch_hgnc_ids.map { _meta, file -> file }
        )
        ch_versions = ch_versions.mix(ENSEMBLVEP_FILTERVEP.out.versions)

        ch_vcf = ENSEMBLVEP_FILTERVEP.out.output
    }

    BCFTOOLS_VIEW (
        ch_vcf.map { meta, vcf -> [ meta, vcf, [] ] },
        [],
        [],
        []
    )
    ch_versions = ch_versions.mix(BCFTOOLS_VIEW.out.versions)

    emit:
    vcf      = BCFTOOLS_VIEW.out.vcf // channel: [ val(meta), path(vcf) ]
    tbi      = BCFTOOLS_VIEW.out.tbi // channel: [ val(meta), path(tbi) ]
    versions = ch_versions           // channel: [ path(versions.yml) ]
}

