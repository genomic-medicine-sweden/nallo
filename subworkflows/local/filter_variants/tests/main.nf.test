nextflow_workflow {

    name "Test Workflow FILTER_VARIANTS"
    script "../main.nf"
    config "./nextflow.config"
    workflow "FILTER_VARIANTS"

    setup {
        run("GUNZIP") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id:'hg38' ],
                    file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                ]
                """
            }
        }
        run("SAMTOOLS_FAIDX") {
            script "../../../../modules/nf-core/samtools/faidx/main.nf"
            process {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = [[],[]]
                """
            }
        }

        run("MINIMAP2_ALIGN") {
            script "../../../../modules/nf-core/minimap2/align/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'test', num_intervals:1 ],
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.fastq.gz', checkIfExists: true)
                ]
                input[1] = GUNZIP.out.gunzip
                input[2] = true
                input[3] = 'csi'
                input[4] = false
                input[5] = false
                """
            }
        }

        run("SHORT_VARIANT_CALLING") {
            script "../../short_variant_calling/main.nf"
            process {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                    .join(MINIMAP2_ALIGN.out.index)
                    .join(Channel.of([
                        [ id: 'test', num_intervals:1 ],
                            file(params.pipelines_testdata_base_path + 'reference/test_data.bed', checkifexists: true)
                    ]))
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = [[],[]]
                input[4] = [[],[]]
            """
            }
        }
        run("UNTAR") {
            script "../../../../modules/nf-core/untar/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'vep_cache' ],
                    file(params.pipelines_testdata_base_path + 'reference/vep_cache_test_data.tar.gz', checkIfExists:true)
                ]
                """
            }
        }
        run("SNV_ANNOTATION") {
            script "../../snv_annotation/main.nf"
            process {
                """
                input[0] = SHORT_VARIANT_CALLING.out.combined_bcf
                input[1] = [
                    file(params.pipelines_testdata_base_path + 'reference/cadd.v1.6.hg38.test_data.zip', checkIfExists: true)
                ]
                input[2] = GUNZIP.out.gunzip
                input[3] = SAMTOOLS_FAIDX.out.fai
                input[4] = UNTAR.out.untar.map { meta, cache -> cache}
                input[5] = Channel.value('110')
                input[6] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ]).splitCsv(header:true).map { row -> row.vep_files }.collect()
                input[7] = false
                input[8] = Channel.value([])
                input[9] = null
                input[10] = null
            """
            }
        }
    }

    test("vcf, [[],[]], false") {

        when {
            workflow {
                """
                input[0] = SNV_ANNOTATION.out.vcf
                input[1] = Channel.of([[],[]])
                input[2] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    file(workflow.out.tbi.get(0).get(1)).name,
                    path(workflow.out.vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.vcf.get(0).get(1)).vcf.summary,
                ).match() }
            )
        }
    }

    test("vcf, hgnc_ids, true") {
        tag "hgnc"
        when {
            workflow {
                """
                input[0] = SNV_ANNOTATION.out.vcf
                input[1] = Channel.of('HGNC:4826')
                    .collectFile(name: 'hgnc_ids.txt')
                    .map { file -> [ [ id: 'hgnc_ids' ], file ] }
                input[2] = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    file(workflow.out.tbi.get(0).get(1)).name,
                    path(workflow.out.vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.vcf.get(0).get(1)).vcf.summary,
                ).match() }
            )
        }
    }
}
