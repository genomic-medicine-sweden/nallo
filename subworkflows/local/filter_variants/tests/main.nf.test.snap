{
    "vcf, [[],[]], false": {
        "content": [
            [
                "versions.yml:md5,7d9185c65ba6ef50b36f1ffd53928cfe"
            ],
            "test_filtered.vcf.gz.tbi",
            "VcfFile [chromosomes=[chrX, chr16, chr17], sampleCount=3, variantCount=202, phased=false, phasedAutodetect=false]"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:02:16.062233877"
    },
    "vcf, hgnc_ids, true": {
        "content": [
            [
                "versions.yml:md5,11806a211f1ea5f7f0658ca263c72b84",
                "versions.yml:md5,7d9185c65ba6ef50b36f1ffd53928cfe"
            ],
            "test_filtered.vcf.gz.tbi",
            "VcfFile [chromosomes=[chr16], sampleCount=3, variantCount=86, phased=false, phasedAutodetect=false]"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:03:03.034454164"
    }
}